Created Car model, DAO and Service for working with it.

zaitsev-serhei opened the pull request:
Updated SQL statements for queries.
Implemented methods for Car Service as per requirements
Maven passed

zaitsev-serhei committed:
Created Car.class and DAO, Service for working with it.
Updated SQL statements for queries.
Implemented methods for Car Service as per requirements
Maven passed

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,200 @@
+    @Override
+    public boolean delete(Long id) {
+        String query = "UPDATE cars SET isDeleted = true "
+                + "WHERE id = ? AND isDeleted = false;";
```
Не вижу смысла второго условия
```suggestion
        String query = "UPDATE cars SET isDeleted = true "
                + "WHERE id = ?;";
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,200 @@
+        return driverSet;
+    }
+
+    private int removeDriversForCar(Car car) {
```
```suggestion
    private void removeDriversForCar(Car car) {
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,200 @@
+
+    private Set<Driver> updateDriversForCar(Car car) {
+        Set<Driver> driverSet = car.getDriverSet();
+        if (!driverSet.isEmpty()) {
```
if (!driverSet.isEmpty()) - это лишнее

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,200 @@
+            resultSet.next();
+            Long recordId = resultSet.getObject(1, Long.class);
+            car.setId(recordId);
+            removeDriversForCar(car);
+            car.setDriverSet(updateDriversForCar(car));
```
![image](https://github.com/zaitsev-serhei/taxi-service/assets/20326858/c10a2245-0324-4aee-985a-cdc219a6d5ff)

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,200 @@
+        this.connectionUtil = connectionUtil;
+    }
+
+    @Transactional
```
Ты протестировал, или это работает?

zaitsev-serhei replied:
Создание машины работает. Дополнительных параметров к аннотации не потребовалось, как оказалось

Rommelua replied:
То что авто создается - это понятно. А есть ли транзакция?

Rommelua commented on src/main/java/com/taxi/model/Car.java:
```diff
@@ -0,0 +1,72 @@
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+    private HashSet<Driver> driverSet;
```
Всегда используем интерфейсы
```suggestion
    private Set<Driver> drivers;
```

zaitsev-serhei committed:
updated SQL queries for CarDaoImpl adding Drivers to a Car
added overrides to Driver.class equals and hashCode to add values to a Set properly
refactored the code as per requirements

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+            resultSet.next();
+            Long recordId = resultSet.getObject(1, Long.class);
+            car.setId(recordId);
+            removeDriversForCar(car, connection);
```
Если мы только создаем машину, то откуда возьмутся кандидаты на ремув?))

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+        return driver;
+    }
+
+    private Set<Driver> updateDriversForCar(Car car, Connection connection) {
```
Зачем возвращать сет, который мы просто берем у машины и ничего с ним не делаем?
```suggestion
    private void addDriversForCar(Car car, Connection connection) {
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+        }
+    }
+
+    private Set<Driver> getDriverSet(Car car, Connection connection) {
```
```suggestion
    private Set<Driver> getDrivers(Car car, Connection connection) {
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+        String removeDriversString = "SELECT d.id, d.name, d.licence, d.isDeleted"
+                + " FROM drivers  d"
+                + " JOIN cars_drivers cd ON d.id = cd.driver_id "
+                + " WHERE car_id = ?";
```
is_deleted

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+        }
+    }
+
+    private Car getCar(ResultSet resultSet) throws SQLException {
```
Предлагаю эти методы назвать **parse**Car, **parse**Manufacturer - чтобы не путать с **get** из БД.

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+            while (resultSet.next()) {
+                carList.add(getCar(resultSet));
+            }
+            return carList;
```
драйверы присутствуют?

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+            preparedStatement.setLong(3, car.getId());
+            preparedStatement.executeUpdate();
+            removeDriversForCar(car, connection);
+            car.setDriverSet(updateDriversForCar(car, connection));
```
car.setDriverSet - в чем смысл этого действия?

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+                car = getCar(resultSet);
+            }
+            car.setDriverSet(getDriverSet(car,connection));
+            return Optional.ofNullable(car);
```
ofNullable - в каком случае car будет null?

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,229 @@
+            Long recordId = resultSet.getObject(1, Long.class);
+            car.setId(recordId);
+            removeDriversForCar(car, connection);
+            car.setDriverSet(updateDriversForCar(car, connection));
```
car.setDriverSet - в чем смысл этого действия?

Rommelua commented on src/main/java/com/taxi/model/Car.java:
```diff
@@ -0,0 +1,72 @@
+    private Long id;
+    private String model;
+    private Manufacturer manufacturer;
+    private Set<Driver> driverSet;
```
```suggestion
    private Set<Driver> drivers;
```

zaitsev-serhei committed:
Update src/main/java/com/taxi/model/Car.java

Co-authored-by: Leonid Sivko <rommel@ukr.net>

zaitsev-serhei committed:
refactored the code regarding comments in PR #4

zaitsev-serhei committed:
Update Car.java

zaitsev-serhei committed:
refactored the code regarding comments in PR #4

zaitsev-serhei committed:
Merge remote-tracking branch 'origin/add_car' into add_car

# Conflicts:
#	src/main/java/com/taxi/model/Car.java

zaitsev-serhei committed:
refactored the code regarding comments in PR #4

zaitsev-serhei committed:
added transactions to the DAO layers.
tested for division by 0 for CarDaoImpl
Checked Maven

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,230 @@
+    }
+
+    @Override
+    public Car create(Car car) {
```
А как же драйверы?
И не вижу ролбек транзакции
![image](https://github.com/zaitsev-serhei/taxi-service/assets/20326858/b4c724b9-6b0b-4f1a-b5fd-11a9b0c8021e)

Rommelua replied:
Вопрос остается - А как же драйверы? почему в этом методе ты не сохраняешь связи, которые потенциально могут быть у этой машины?

zaitsev-serhei replied:
добавил в метод вызов приватного метода по апдейту связей с водителями. Понял суть вопроса

zaitsev-serhei committed:
updated Dao methods create and update for each model
added transaction rollback functionality
refactored the code according to the requirements
Maven tests passed

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,257 @@
+        try (PreparedStatement preparedStatement =
+                     connection.prepareStatement(updateStatement)) {
+            connection.setAutoCommit(false);
+            preparedStatement.setString(1, car.getModel());
+            preparedStatement.setLong(2, car.getManufacturer().getId());
+            preparedStatement.setLong(3, car.getId());
+            preparedStatement.executeUpdate();
```
Предлагаю рефакторинг. Делаем еще один приватный метод. И у нас будет 3 красивых вызова подряд приватных методов. Главное его хорошо назвать. 
![image](https://github.com/zaitsev-serhei/taxi-service/assets/20326858/05a2a7cc-0a61-44d5-a6a3-f4f922d87393)

zaitsev-serhei committed:
refactored the Update method for CarDaoImpl

zaitsev-serhei committed:
refactored the Update method for CarDaoImpl

Rommelua commented on src/main/java/com/taxi/config/AppConfig.java:
```diff
@@ -14,4 +14,5 @@ public class AppConfig {
     public AppConfig(Environment environment) {
         this.environment = environment;
     }
+
```
Ненужное изменение)

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,267 @@
+        Connection connection = connectionUtil.getConnection();
+        try (PreparedStatement preparedStatement = connection.prepareStatement(createStatement,
+                Statement.RETURN_GENERATED_KEYS)) {
+            connection.setAutoCommit(false);
+            preparedStatement.setString(1, (car.getModel()));
+            preparedStatement.setLong(2, car.getManufacturer().getId());
+            preparedStatement.executeUpdate();
+            addDriversForCar(car,connection);
+            connection.commit();
```
Давай сделаем по аналогии с Апдейтом
```suggestion
            connection.setAutoCommit(false);
            addCar(car,connection);;
            addDriversForCar(car,connection);
            connection.commit();
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,267 @@
+        } catch (Exception e) {
+            if (connection != null) {
+                try {
+                    System.out.println("Transaction in Update Car was rolled back ");
```
```suggestion
```

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,267 @@
+                    System.out.println("Transaction in Update Car was rolled back ");
+                    connection.rollback();
+                } catch (SQLException exception) {
+                    exception.printStackTrace();
```
Везде где есть printStackTrace добавь TODO заменить на логгер. Потом Логгер добавим в проект.

Rommelua commented on src/main/java/com/taxi/dao/impl/CarDaoImpl.java:
```diff
@@ -0,0 +1,267 @@
+        } catch (Exception exception) {
+            throw new DataProcessingException("Can`t invoke update Car "
+                    + car, exception);
+        }
```
В этих методах catch не нужен. Пробрасывай выше как в parse методах. Ведь вызовы идут внутри трай блоков.

Rommelua commented on src/main/java/com/taxi/dao/impl/DriverDaoImpl.java:
```diff
@@ -82,17 +84,33 @@ public List<Driver> getAll() {
+                    System.out.println("Transaction in Update Driver was rolled back ");
+                    connection.rollback();
+                } catch (SQLException exception) {
+                    exception.printStackTrace();
```
И тут тудушки для printStackTrace тоже

zaitsev-serhei committed:
added TODO to DAO methods
updated CarDaoImpl methods as per comments in PR

zaitsev-serhei committed:
added TODO to DAO methods
updated CarDaoImpl methods as per comments in PR

zaitsev-serhei committed:
refactored the line )))
